File: packages/drawingml/src/theme.ts
/**
 * Theme parsing — colors and fonts from a DrawingML theme part
 * (e.g. word/theme/theme1.xml, ppt/theme/theme1.xml).
 *
 * Extracts color scheme (dk1, lt1, dk2, lt2, accent1-6, hlink, folHlink)
 * and font scheme (majorFont, minorFont) from the theme.
 *
 * OOXML Reference:
 * - Uses DrawingML namespace (a:)
 * - Colors can be srgbClr, sysClr, or schemeClr
 */

import { DEFAULT_THEME_COLORS } from './color';
import {
  findChild,
  findChildren,
  getAttribute,
  getChildElements,
  getLocalName,
  type XmlLike,
} from './xml';

/**
 * Theme color scheme (a:clrScheme)
 *
 * @public
 */
export interface ThemeColorScheme {
  /** Dark 1 color (usually black) */
  dk1?: string;
  /** Light 1 color (usually white) */
  lt1?: string;
  /** Dark 2 color */
  dk2?: string;
  /** Light 2 color */
  lt2?: string;
  /** Accent colors 1-6 */
  accent1?: string;
  accent2?: string;
  accent3?: string;
  accent4?: string;
  accent5?: string;
  accent6?: string;
  /** Hyperlink color */
  hlink?: string;
  /** Followed hyperlink color */
  folHlink?: string;
}

/**
 * Theme font (with script variants)
 *
 * @public
 */
export interface ThemeFont {
  /** Latin font */
  latin?: string;
  /** East Asian font */
  ea?: string;
  /** Complex script font */
  cs?: string;
  /** Script-specific fonts */
  fonts?: Record<string, string>;
}

/**
 * Theme font scheme (a:fontScheme)
 *
 * @public
 */
export interface ThemeFontScheme {
  /** Major font (headings) */
  majorFont?: ThemeFont;
  /** Minor font (body text) */
  minorFont?: ThemeFont;
}

/**
 * Theme (from a theme1.xml part)
 *
 * @public
 */
export interface Theme {
  /** Theme name */
  name?: string;
  /** Color scheme */
  colorScheme?: ThemeColorScheme;
  /** Font scheme */
  fontScheme?: ThemeFontScheme;
  /** Format scheme (fills, lines, effects) - simplified */
  formatScheme?: {
    name?: string;
  };
}

/**
 * Default font scheme
 */
const DEFAULT_FONTS: ThemeFontScheme = {
  majorFont: {
    latin: 'Calibri Light',
    ea: '',
    cs: '',
    fonts: {},
  },
  minorFont: {
    latin: 'Calibri',
    ea: '',
    cs: '',
    fonts: {},
  },
};

/**
 * Default theme when no theme part exists
 */
const DEFAULT_THEME: Theme = {
  name: 'Office Theme',
  colorScheme: DEFAULT_THEME_COLORS,
  fontScheme: DEFAULT_FONTS,
};

/**
 * Color slot names in theme
 */
const COLOR_SLOTS = [
  'dk1',
  'lt1',
  'dk2',
  'lt2',
  'accent1',
  'accent2',
  'accent3',
  'accent4',
  'accent5',
  'accent6',
  'hlink',
  'folHlink',
] as const;

/**
 * Parse a color element (srgbClr, sysClr, or schemeClr)
 *
 * @param element - Color child element
 * @returns Hex color value (6 characters, no #)
 */
function parseThemeColorElement(element: XmlLike | null): string | null {
  if (!element) return null;

  const localName = getLocalName(element.name || '');

  switch (localName) {
    case 'srgbClr': {
      // Direct RGB color: <a:srgbClr val="4472C4"/>
      const val = getAttribute(element, 'a', 'val') ?? getAttribute(element, null, 'val');
      return val ?? null;
    }

    case 'sysClr': {
      // System color with fallback: <a:sysClr val="windowText" lastClr="000000"/>
      // Use lastClr as the fallback since we can't access actual system colors
      const lastClr =
        getAttribute(element, 'a', 'lastClr') ?? getAttribute(element, null, 'lastClr');
      if (lastClr) return lastClr;

      // Fallback based on common system color names
      const val = getAttribute(element, 'a', 'val') ?? getAttribute(element, null, 'val');
      switch (val) {
        case 'windowText':
        case 'menuText':
        case 'captionText':
        case 'btnText':
          return '000000';
        case 'window':
        case 'menu':
        case 'btnFace':
        case 'btnHighlight':
          return 'FFFFFF';
        case 'highlight':
          return '0078D7';
        case 'highlightText':
          return 'FFFFFF';
        case 'grayText':
          return '808080';
        default:
          return null;
      }
    }

    case 'schemeClr': {
      // Reference to another scheme color - rare in color scheme itself
      // Usually found in fill/line definitions with modifiers
      // For the color scheme, we just need the val
      const val = getAttribute(element, 'a', 'val') ?? getAttribute(element, null, 'val');
      // This is a reference, not a final color - return null for now
      // The actual resolution would need the full color scheme
      return val === 'phClr' ? null : null;
    }

    default:
      return null;
  }
}

/**
 * Parse the color scheme from a:clrScheme element
 *
 * @param clrScheme - The a:clrScheme element
 * @returns ThemeColorScheme with resolved hex colors
 */
function parseColorScheme(clrScheme: XmlLike | null): ThemeColorScheme {
  const result: ThemeColorScheme = { ...DEFAULT_THEME_COLORS };

  if (!clrScheme) return result;

  // Each color slot has a child element with the slot name
  for (const slot of COLOR_SLOTS) {
    // Find the slot element (e.g., a:dk1, a:accent1)
    const slotElement = findChild(clrScheme, 'a', slot);

    if (slotElement) {
      // The actual color is a child (srgbClr, sysClr, or schemeClr)
      const children = getChildElements(slotElement);
      if (children.length > 0) {
        const color = parseThemeColorElement(children[0]);
        if (color) {
          result[slot] = color;
        }
      }
    }
  }

  return result;
}

/**
 * Parse a font definition (majorFont or minorFont)
 *
 * @param fontElement - The a:majorFont or a:minorFont element
 * @returns ThemeFont with font family names
 */
function parseThemeFonts(fontElement: XmlLike | null): ThemeFont {
  const result: ThemeFont = {
    latin: '',
    ea: '',
    cs: '',
    fonts: {},
  };

  if (!fontElement) return result;

  // Parse main font elements
  const latinEl = findChild(fontElement, 'a', 'latin');
  if (latinEl) {
    result.latin =
      getAttribute(latinEl, 'a', 'typeface') ?? getAttribute(latinEl, null, 'typeface') ?? '';
  }

  const eaEl = findChild(fontElement, 'a', 'ea');
  if (eaEl) {
    result.ea = getAttribute(eaEl, 'a', 'typeface') ?? getAttribute(eaEl, null, 'typeface') ?? '';
  }

  const csEl = findChild(fontElement, 'a', 'cs');
  if (csEl) {
    result.cs = getAttribute(csEl, 'a', 'typeface') ?? getAttribute(csEl, null, 'typeface') ?? '';
  }

  // Parse script-specific fonts (a:font elements with script attribute)
  const fontElements = findChildren(fontElement, 'a', 'font');
  for (const font of fontElements) {
    const script = getAttribute(font, 'a', 'script') ?? getAttribute(font, null, 'script');
    const typeface = getAttribute(font, 'a', 'typeface') ?? getAttribute(font, null, 'typeface');

    if (script && typeface) {
      result.fonts = result.fonts || {};
      result.fonts[script] = typeface;
    }
  }

  return result;
}

/**
 * Parse the font scheme from a:fontScheme element
 *
 * @param fontScheme - The a:fontScheme element
 * @returns ThemeFontScheme with major and minor fonts
 */
function parseFontScheme(fontScheme: XmlLike | null): ThemeFontScheme {
  const result: ThemeFontScheme = { ...DEFAULT_FONTS };

  if (!fontScheme) return result;

  const majorFontEl = findChild(fontScheme, 'a', 'majorFont');
  if (majorFontEl) {
    result.majorFont = parseThemeFonts(majorFontEl);
  }

  const minorFontEl = findChild(fontScheme, 'a', 'minorFont');
  if (minorFontEl) {
    result.minorFont = parseThemeFonts(minorFontEl);
  }

  return result;
}

/**
 * Parse a theme part from its root `a:theme` element.
 *
 * The host is responsible for XML parsing; pass the parsed root element
 * (any {@link XmlLike}-compatible tree).
 *
 * @param root - Root a:theme element, or null/undefined if not present
 * @returns Parsed Theme object with colors and fonts
 *
 * @public
 */
export function parseTheme(root: XmlLike | null | undefined): Theme {
  // Return defaults if no theme element
  if (!root) {
    return { ...DEFAULT_THEME };
  }

  try {
    // Get theme name from root element
    const themeName =
      getAttribute(root, 'a', 'name') ?? getAttribute(root, null, 'name') ?? 'Office Theme';

    // Find a:themeElements which contains clrScheme and fontScheme
    const themeElements = findChild(root, 'a', 'themeElements');

    // Parse color scheme
    const clrScheme = findChild(themeElements, 'a', 'clrScheme');
    const colorScheme = parseColorScheme(clrScheme);

    // Parse font scheme
    const fontSchemeEl = findChild(themeElements, 'a', 'fontScheme');
    const fontScheme = parseFontScheme(fontSchemeEl);

    return {
      name: themeName,
      colorScheme,
      fontScheme,
    };
  } catch (error) {
    console.warn('Failed to parse theme:', error);
    return { ...DEFAULT_THEME };
  }
}

/**
 * Map a `w:themeFontLang` EastAsian language tag to the DrawingML `script`
 * attribute used by the theme's `<a:font script="…">` entries.
 */
function eastAsiaLangToScript(lang: string): string | null {
  const l = lang.toLowerCase();
  if (l.startsWith('ja')) return 'Jpan';
  if (l.startsWith('ko')) return 'Hang';
  if (l.startsWith('zh')) {
    // Traditional Chinese locales use the Hant typeface; everything else Hans.
    if (l.includes('hant') || /-(tw|hk|mo)\b/.test(l)) return 'Hant';
    return 'Hans';
  }
  return null;
}

/**
 * Map a `w:themeFontLang` bidi/complex-script language tag to a DrawingML
 * `script` attribute.
 */
function bidiLangToScript(lang: string): string | null {
  const l = lang.toLowerCase();
  if (l.startsWith('ar')) return 'Arab';
  if (l.startsWith('he') || l.startsWith('iw')) return 'Hebr';
  if (l.startsWith('th')) return 'Thai';
  if (l.startsWith('hi') || l.startsWith('mr') || l.startsWith('ne')) return 'Deva';
  return null;
}

/**
 * Resolve empty EastAsian/complex-script theme font slots using the document's
 * `w:themeFontLang`.
 *
 * Office's default theme leaves `<a:ea typeface="">`/`<a:cs typeface="">` empty
 * and instead lists the real typeface per script (`<a:font script="Jpan" …>`).
 * Word selects among those by the user/document language declared in
 * `w:themeFontLang`. We mirror that here so a run that references
 * `minorEastAsia` in a Japanese document resolves to `ＭＳ 明朝` rather than an
 * empty string — an empty font breaks line-break measurement (canvas falls back
 * to its 10px default) while the painter renders CJK glyphs at full width, so
 * text overflows the right margin.
 *
 * Mutates the theme in place. This is safe for round-tripping because the
 * theme part is preserved verbatim on save; the parsed `Theme` is only
 * consumed for layout/rendering.
 *
 * @public
 */
export function applyThemeFontLang(
  theme: Theme | null | undefined,
  themeFontLang: { eastAsia?: string; bidi?: string } | undefined
): void {
  if (!theme?.fontScheme || !themeFontLang) return;

  const eaScript = themeFontLang.eastAsia ? eastAsiaLangToScript(themeFontLang.eastAsia) : null;
  const csScript = themeFontLang.bidi ? bidiLangToScript(themeFontLang.bidi) : null;

  for (const font of [theme.fontScheme.majorFont, theme.fontScheme.minorFont]) {
    if (!font) continue;
    if (!font.ea && eaScript && font.fonts?.[eaScript]) {
      font.ea = font.fonts[eaScript];
    }
    if (!font.cs && csScript && font.fonts?.[csScript]) {
      font.cs = font.fonts[csScript];
    }
  }
}

/**
 * Get a color from the theme by slot name
 *
 * @param theme - Parsed theme
 * @param slot - Color slot name (dk1, lt1, accent1, etc.)
 * @returns Hex color value (6 characters, no #)
 *
 * @public
 */
export function getThemeColor(
  theme: Theme | null | undefined,
  slot: keyof ThemeColorScheme
): string {
  if (!theme?.colorScheme) {
    return DEFAULT_THEME_COLORS[slot] ?? '000000';
  }

  return theme.colorScheme[slot] ?? DEFAULT_THEME_COLORS[slot] ?? '000000';
}

/**
 * Get the major font (heading font) from theme
 *
 * @param theme - Parsed theme
 * @param script - Optional script code (defaults to latin)
 * @returns Font family name
 *
 * @public
 */
export function getMajorFont(theme: Theme | null | undefined, script: string = 'latin'): string {
  if (!theme?.fontScheme?.majorFont) {
    return DEFAULT_FONTS.majorFont?.latin ?? 'Calibri Light';
  }

  const majorFont = theme.fontScheme.majorFont;

  if (script === 'latin') return majorFont.latin || 'Calibri Light';
  if (script === 'ea') return majorFont.ea || '';
  if (script === 'cs') return majorFont.cs || '';

  // Check script-specific fonts
  if (majorFont.fonts?.[script]) {
    return majorFont.fonts[script];
  }

  // Default to latin
  return majorFont.latin || 'Calibri Light';
}

/**
 * Get the minor font (body font) from theme
 *
 * @param theme - Parsed theme
 * @param script - Optional script code (defaults to latin)
 * @returns Font family name
 *
 * @public
 */
export function getMinorFont(theme: Theme | null | undefined, script: string = 'latin'): string {
  if (!theme?.fontScheme?.minorFont) {
    return DEFAULT_FONTS.minorFont?.latin ?? 'Calibri';
  }

  const minorFont = theme.fontScheme.minorFont;

  if (script === 'latin') return minorFont.latin || 'Calibri';
  if (script === 'ea') return minorFont.ea || '';
  if (script === 'cs') return minorFont.cs || '';

  // Check script-specific fonts
  if (minorFont.fonts?.[script]) {
    return minorFont.fonts[script];
  }

  // Default to latin
  return minorFont.latin || 'Calibri';
}

/**
 * Resolve a theme font reference to an actual font name
 *
 * Theme font references are like: majorAscii, majorHAnsi, minorAscii, minorHAnsi, etc.
 *
 * @param theme - Parsed theme
 * @param themeRef - Theme font reference
 * @returns Font family name
 *
 * @public
 */
export function resolveThemeFontRef(theme: Theme | null | undefined, themeRef: string): string {
  if (!themeRef) return 'Calibri';

  // Parse the reference: major/minor + script type
  const isMajor = themeRef.toLowerCase().includes('major');
  const isMinor = themeRef.toLowerCase().includes('minor');

  // Determine script from reference
  let script = 'latin';
  const lowerRef = themeRef.toLowerCase();

  if (lowerRef.includes('eastasia')) {
    script = 'ea';
  } else if (lowerRef.includes('bidi') || lowerRef.includes('cs')) {
    script = 'cs';
  }
  // ascii and hAnsi both map to latin

  if (isMajor) {
    return getMajorFont(theme, script);
  } else if (isMinor) {
    return getMinorFont(theme, script);
  }

  // Default to minor latin
  return getMinorFont(theme, 'latin');
}

/**
 * Get all font families from the theme for preloading
 *
 * @param theme - Parsed theme
 * @returns Array of unique font family names
 *
 * @public
 */
export function getThemeFonts(theme: Theme | null | undefined): string[] {
  const fonts = new Set<string>();

  if (theme?.fontScheme) {
    const { majorFont, minorFont } = theme.fontScheme;

    // Add main fonts
    if (majorFont?.latin) fonts.add(majorFont.latin);
    if (majorFont?.ea) fonts.add(majorFont.ea);
    if (majorFont?.cs) fonts.add(majorFont.cs);

    if (minorFont?.latin) fonts.add(minorFont.latin);
    if (minorFont?.ea) fonts.add(minorFont.ea);
    if (minorFont?.cs) fonts.add(minorFont.cs);

    // Add script-specific fonts
    if (majorFont?.fonts) {
      for (const font of Object.values(majorFont.fonts)) {
        if (font) fonts.add(font);
      }
    }

    if (minorFont?.fonts) {
      for (const font of Object.values(minorFont.fonts)) {
        if (font) fonts.add(font);
      }
    }
  }

  // Remove empty strings
  fonts.delete('');

  return Array.from(fonts);
}

/**
 * Get the default theme (Office 2016 theme)
 *
 * @returns Default Theme object
 *
 * @public
 */
export function getDefaultTheme(): Theme {
  return { ...DEFAULT_THEME };
}

File: crates/docx-parse/src/theme.rs
//! DrawingML theme color/font subset required by S2 and future style resolution.

use indexmap::{IndexMap, IndexSet};
use serde::{Deserialize, Serialize};

use crate::settings::ThemeFontLanguage;
use crate::settings::incumbent_utf8_text_boundary;
use crate::xml::{ParseBudget, ParseError, XmlElement, parse_xml};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThemeColorScheme {
    pub dk1: String,
    pub lt1: String,
    pub dk2: String,
    pub lt2: String,
    pub accent1: String,
    pub accent2: String,
    pub accent3: String,
    pub accent4: String,
    pub accent5: String,
    pub accent6: String,
    pub hlink: String,
    pub fol_hlink: String,
}

impl Default for ThemeColorScheme {
    fn default() -> Self {
        Self {
            dk1: "000000".to_owned(),
            lt1: "FFFFFF".to_owned(),
            dk2: "44546A".to_owned(),
            lt2: "E7E6E6".to_owned(),
            accent1: "4472C4".to_owned(),
            accent2: "ED7D31".to_owned(),
            accent3: "A5A5A5".to_owned(),
            accent4: "FFC000".to_owned(),
            accent5: "5B9BD5".to_owned(),
            accent6: "70AD47".to_owned(),
            hlink: "0563C1".to_owned(),
            fol_hlink: "954F72".to_owned(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ThemeFont {
    pub latin: String,
    pub ea: String,
    pub cs: String,
    pub fonts: IndexMap<String, String>,
}

impl ThemeFont {
    fn default_major() -> Self {
        Self {
            latin: "Calibri Light".to_owned(),
            ea: String::new(),
            cs: String::new(),
            fonts: IndexMap::new(),
        }
    }

    fn default_minor() -> Self {
        Self {
            latin: "Calibri".to_owned(),
            ea: String::new(),
            cs: String::new(),
            fonts: IndexMap::new(),
        }
    }

    fn empty() -> Self {
        Self {
            latin: String::new(),
            ea: String::new(),
            cs: String::new(),
            fonts: IndexMap::new(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThemeFontScheme {
    pub major_font: ThemeFont,
    pub minor_font: ThemeFont,
}

impl Default for ThemeFontScheme {
    fn default() -> Self {
        Self {
            major_font: ThemeFont::default_major(),
            minor_font: ThemeFont::default_minor(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Theme {
    pub name: String,
    pub color_scheme: ThemeColorScheme,
    pub font_scheme: ThemeFontScheme,
}

impl Default for Theme {
    fn default() -> Self {
        Self {
            name: "Office Theme".to_owned(),
            color_scheme: ThemeColorScheme::default(),
            font_scheme: ThemeFontScheme::default(),
        }
    }
}

pub fn parse_theme(
    xml: Option<&[u8]>,
    part: &str,
    budget: &mut ParseBudget<'_>,
) -> Result<Theme, ParseError> {
    let Some(xml) = xml.filter(|value| !is_blank(value)) else {
        return Ok(Theme::default());
    };
    if !incumbent_utf8_text_boundary(xml) {
        return Ok(Theme::default());
    }
    let document = parse_xml(xml, part, budget)?;
    Ok(parse_theme_element(document.root()))
}

pub fn parse_theme_element(root: Option<&XmlElement>) -> Theme {
    let Some(root) = root else {
        return Theme::default();
    };
    let theme_elements = root.child("a", "themeElements");
    Theme {
        name: root
            .attribute(Some("a"), "name")
            .unwrap_or("Office Theme")
            .to_owned(),
        color_scheme: parse_color_scheme(
            theme_elements.and_then(|element| element.child("a", "clrScheme")),
        ),
        font_scheme: parse_font_scheme(
            theme_elements.and_then(|element| element.child("a", "fontScheme")),
        ),
    }
}

fn parse_color_scheme(element: Option<&XmlElement>) -> ThemeColorScheme {
    let mut colors = ThemeColorScheme::default();
    let Some(element) = element else {
        return colors;
    };
    for slot in [
        "dk1", "lt1", "dk2", "lt2", "accent1", "accent2", "accent3", "accent4", "accent5",
        "accent6", "hlink", "folHlink",
    ] {
        let value = element
            .child("a", slot)
            .and_then(|slot| slot.child_elements().next())
            .and_then(parse_theme_color_element);
        if let Some(value) = value {
            colors.set(slot, value);
        }
    }
    colors
}

fn parse_theme_color_element(element: &XmlElement) -> Option<String> {
    match element.local_name() {
        "srgbClr" => nonempty(element.attribute(Some("a"), "val")),
        "sysClr" => {
            if let Some(last_color) = nonempty(element.attribute(Some("a"), "lastClr")) {
                return Some(last_color);
            }
            match element.attribute(Some("a"), "val") {
                Some("windowText" | "menuText" | "captionText" | "btnText") => {
                    Some("000000".to_owned())
                }
                Some("window" | "menu" | "btnFace" | "btnHighlight" | "highlightText") => {
                    Some("FFFFFF".to_owned())
                }
                Some("highlight") => Some("0078D7".to_owned()),
                Some("grayText") => Some("808080".to_owned()),
                _ => None,
            }
        }
        // The incumbent deliberately leaves scheme references unresolved when
        // they occur inside the scheme itself.
        "schemeClr" => None,
        _ => None,
    }
}

fn parse_font_scheme(element: Option<&XmlElement>) -> ThemeFontScheme {
    let Some(element) = element else {
        return ThemeFontScheme::default();
    };
    ThemeFontScheme {
        major_font: element
            .child("a", "majorFont")
            .map(parse_theme_font)
            .unwrap_or_else(ThemeFont::default_major),
        minor_font: element
            .child("a", "minorFont")
            .map(parse_theme_font)
            .unwrap_or_else(ThemeFont::default_minor),
    }
}

fn parse_theme_font(element: &XmlElement) -> ThemeFont {
    let mut font = ThemeFont::empty();
    font.latin = typeface(element.child("a", "latin"));
    font.ea = typeface(element.child("a", "ea"));
    font.cs = typeface(element.child("a", "cs"));
    for script_font in element.children_named("a", "font") {
        let script = nonempty(script_font.attribute(Some("a"), "script"));
        let typeface = nonempty(script_font.attribute(Some("a"), "typeface"));
        if let (Some(script), Some(typeface)) = (script, typeface) {
            font.fonts.insert(script, typeface);
        }
    }
    font
}

fn typeface(element: Option<&XmlElement>) -> String {
    element
        .and_then(|element| element.attribute(Some("a"), "typeface"))
        .unwrap_or_default()
        .to_owned()
}

impl ThemeColorScheme {
    fn set(&mut self, slot: &str, value: String) {
        match slot {
            "dk1" => self.dk1 = value,
            "lt1" => self.lt1 = value,
            "dk2" => self.dk2 = value,
            "lt2" => self.lt2 = value,
            "accent1" => self.accent1 = value,
            "accent2" => self.accent2 = value,
            "accent3" => self.accent3 = value,
            "accent4" => self.accent4 = value,
            "accent5" => self.accent5 = value,
            "accent6" => self.accent6 = value,
            "hlink" => self.hlink = value,
            "folHlink" => self.fol_hlink = value,
            _ => {}
        }
    }

    fn get(&self, slot: &str) -> Option<&str> {
        match slot {
            "dk1" => Some(&self.dk1),
            "lt1" => Some(&self.lt1),
            "dk2" => Some(&self.dk2),
            "lt2" => Some(&self.lt2),
            "accent1" => Some(&self.accent1),
            "accent2" => Some(&self.accent2),
            "accent3" => Some(&self.accent3),
            "accent4" => Some(&self.accent4),
            "accent5" => Some(&self.accent5),
            "accent6" => Some(&self.accent6),
            "hlink" => Some(&self.hlink),
            "folHlink" => Some(&self.fol_hlink),
            _ => None,
        }
        .map(String::as_str)
    }
}

pub fn apply_theme_font_lang(theme: &mut Theme, language: Option<&ThemeFontLanguage>) {
    let Some(language) = language else { return };
    let east_asia_script = language
        .east_asia
        .as_deref()
        .and_then(east_asia_lang_to_script);
    let bidi_script = language.bidi.as_deref().and_then(bidi_lang_to_script);
    for font in [
        &mut theme.font_scheme.major_font,
        &mut theme.font_scheme.minor_font,
    ] {
        if font.ea.is_empty()
            && let Some(typeface) = east_asia_script.and_then(|script| font.fonts.get(script))
        {
            font.ea.clone_from(typeface);
        }
        if font.cs.is_empty()
            && let Some(typeface) = bidi_script.and_then(|script| font.fonts.get(script))
        {
            font.cs.clone_from(typeface);
        }
    }
}

fn east_asia_lang_to_script(language: &str) -> Option<&'static str> {
    let lower = language.to_ascii_lowercase();
    if lower.starts_with("ja") {
        Some("Jpan")
    } else if lower.starts_with("ko") {
        Some("Hang")
    } else if lower.starts_with("zh") {
        if lower.contains("hant")
            || lower
                .split('-')
                .any(|segment| matches!(segment, "tw" | "hk" | "mo"))
        {
            Some("Hant")
        } else {
            Some("Hans")
        }
    } else {
        None
    }
}

fn bidi_lang_to_script(language: &str) -> Option<&'static str> {
    let lower = language.to_ascii_lowercase();
    if lower.starts_with("ar") {
        Some("Arab")
    } else if lower.starts_with("he") || lower.starts_with("iw") {
        Some("Hebr")
    } else if lower.starts_with("th") {
        Some("Thai")
    } else if lower.starts_with("hi") || lower.starts_with("mr") || lower.starts_with("ne") {
        Some("Deva")
    } else {
        None
    }
}

pub fn get_theme_color(theme: Option<&Theme>, slot: &str) -> String {
    if let Some(value) = theme.and_then(|theme| theme.color_scheme.get(slot)) {
        return value.to_owned();
    }
    let defaults = ThemeColorScheme::default();
    defaults.get(slot).unwrap_or("000000").to_owned()
}

pub fn get_major_font(theme: Option<&Theme>, script: &str) -> String {
    get_font(
        theme.map(|theme| &theme.font_scheme.major_font),
        script,
        "Calibri Light",
    )
}

pub fn get_minor_font(theme: Option<&Theme>, script: &str) -> String {
    get_font(
        theme.map(|theme| &theme.font_scheme.minor_font),
        script,
        "Calibri",
    )
}

fn get_font(font: Option<&ThemeFont>, script: &str, latin_default: &str) -> String {
    let Some(font) = font else {
        return latin_default.to_owned();
    };
    match script {
        "latin" => nonempty(Some(&font.latin)).unwrap_or_else(|| latin_default.to_owned()),
        "ea" => font.ea.clone(),
        "cs" => font.cs.clone(),
        script => font
            .fonts
            .get(script)
            .cloned()
            .or_else(|| nonempty(Some(&font.latin)))
            .unwrap_or_else(|| latin_default.to_owned()),
    }
}

pub fn resolve_theme_font_ref(theme: Option<&Theme>, reference: &str) -> String {
    if reference.is_empty() {
        return "Calibri".to_owned();
    }
    let lower = reference.to_ascii_lowercase();
    let script = if lower.contains("eastasia") {
        "ea"
    } else if lower.contains("bidi") || lower.contains("cs") {
        "cs"
    } else {
        "latin"
    };
    if lower.contains("major") {
        get_major_font(theme, script)
    } else {
        get_minor_font(theme, script)
    }
}

pub fn get_theme_fonts(theme: Option<&Theme>) -> Vec<String> {
    let mut fonts = IndexSet::new();
    if let Some(theme) = theme {
        for font in [&theme.font_scheme.major_font, &theme.font_scheme.minor_font] {
            for value in [&font.latin, &font.ea, &font.cs] {
                if !value.is_empty() {
                    fonts.insert(value.clone());
                }
            }
        }
        for value in theme
            .font_scheme
            .major_font
            .fonts
            .values()
            .chain(theme.font_scheme.minor_font.fonts.values())
        {
            if !value.is_empty() {
                fonts.insert(value.clone());
            }
        }
    }
    fonts.into_iter().collect()
}

pub fn get_default_theme() -> Theme {
    Theme::default()
}

fn nonempty(value: Option<&str>) -> Option<String> {
    value.filter(|value| !value.is_empty()).map(str::to_owned)
}

fn is_blank(value: &[u8]) -> bool {
    std::str::from_utf8(value).is_ok_and(|value| value.trim().is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::xml::ParseLimits;

    fn parse(xml: Option<&str>) -> Theme {
        let limits = ParseLimits::default();
        parse_theme(
            xml.map(str::as_bytes),
            "word/theme/theme1.xml",
            &mut ParseBudget::new(&limits),
        )
        .unwrap()
    }

    #[test]
    fn defaults_for_missing_or_blank_input() {
        assert_eq!(parse(None), Theme::default());
        assert_eq!(parse(Some(" \n\t")), Theme::default());
    }

    #[test]
    fn parses_colors_fonts_system_fallbacks_and_duplicate_scripts() {
        let theme = parse(Some(
            r#"<a:theme name="Golden"><a:themeElements><a:clrScheme><a:dk1><a:sysClr val="windowText"/></a:dk1><a:lt1><a:sysClr val="window" lastClr="FEFEFE"/></a:lt1><a:accent1><a:srgbClr val="BADHEX"/></a:accent1><a:accent2><a:schemeClr val="phClr"/></a:accent2></a:clrScheme><a:fontScheme><a:majorFont><a:latin typeface="Major"/><a:ea typeface=""/><a:cs typeface=""/><a:font script="Jpan" typeface="First"/><a:font script="Jpan" typeface="Last"/></a:majorFont><a:minorFont><a:latin typeface="Minor"/></a:minorFont></a:fontScheme></a:themeElements></a:theme>"#,
        ));
        assert_eq!(theme.name, "Golden");
        assert_eq!(theme.color_scheme.dk1, "000000");
        assert_eq!(theme.color_scheme.lt1, "FEFEFE");
        assert_eq!(theme.color_scheme.accent1, "BADHEX");
        assert_eq!(theme.color_scheme.accent2, "ED7D31");
        assert_eq!(theme.font_scheme.major_font.fonts["Jpan"], "Last");
    }

    #[test]
    fn resolves_language_slots_references_colors_and_unique_font_order() {
        let mut theme = parse(Some(
            r#"<a:theme><a:themeElements><a:fontScheme><a:majorFont><a:latin typeface="Major"/><a:ea typeface=""/><a:cs typeface=""/><a:font script="Hant" typeface="Traditional"/><a:font script="Arab" typeface="Arabic"/></a:majorFont><a:minorFont><a:latin typeface="Minor"/><a:ea typeface=""/><a:cs typeface=""/><a:font script="Hant" typeface="Traditional"/></a:minorFont></a:fontScheme></a:themeElements></a:theme>"#,
        ));
        apply_theme_font_lang(
            &mut theme,
            Some(&ThemeFontLanguage {
                east_asia: Some("zh-TW".to_owned()),
                bidi: Some("ar-SA".to_owned()),
            }),
        );
        assert_eq!(
            resolve_theme_font_ref(Some(&theme), "majorEastAsia"),
            "Traditional"
        );
        assert_eq!(resolve_theme_font_ref(Some(&theme), "minorAscii"), "Minor");
        assert_eq!(get_theme_color(Some(&theme), "missing"), "000000");
        assert_eq!(
            get_theme_fonts(Some(&theme)),
            ["Major", "Traditional", "Arabic", "Minor"]
        );
    }

    #[test]
    fn dtd_and_huge_attributes_are_rejected_without_panics() {
        let limits = ParseLimits::default();
        assert!(matches!(
            parse_theme(
                Some(b"<!DOCTYPE x><a:theme/>"),
                "word/theme/theme1.xml",
                &mut ParseBudget::new(&limits)
            ),
            Err(ParseError::UnsafeXml { .. })
        ));
        let huge = format!(r#"<a:theme name="{}"/>"#, "x".repeat(5_000_000));
        let limits = ParseLimits {
            max_attribute_bytes: 1024,
            ..ParseLimits::default()
        };
        assert!(matches!(
            parse_theme(
                Some(huge.as_bytes()),
                "word/theme/theme1.xml",
                &mut ParseBudget::new(&limits)
            ),
            Err(ParseError::ResourceLimit { .. })
        ));
    }
}
